### ISSUE TITLE
Allow editing of category(body) in admin interface

### ISSUE THREAD

**Issue by dracos** [mysociety/fixmystreet#400]
You should be able to change the category to one of the current categories for the location of the report (and if necessary, it then updates the body too to match).

Also include moving of pin?

**Comment by jonkri** [comment]
Hi, Matthew!

I'm quoting what you wrote on the mailing-list so that it's visible here as well:

<blockquote>A cobrand-specific way of doing this was implemented as part of the Zurich cobrand. Zurich has a heavily customised admin interface (under
templates/web/zurich/admin) and if you look at perllib/FixMyStreet/Cobrand/Zurich.pm, e.g. line 499ish you'll see a bit of code about assigning to a different category. Zurich has fixed categories, and other known features, hence why this was at the time done specifically for that cobrand.</blockquote>

<blockquote>I don't see why it shouldn't be possible to change the category (and therefore the associated body) through the admin, it would just need to be careful that it updated things appropriately. If it's ever needed to be done in the UK, we have manually tweaked the database, as it's been quite rare. This would be a useful feature if you were able to add it :-)</blockquote>

<blockquote>The easiest way for resending would be to then mark whensent as undef, so it was treated as a 'new' report and sent to the new body. This would lose the bit of history that it was sent somewhere else first, some more development would need to be done to store that history if it was needed.</blockquote>

Cheers! :)

**Comment by jonkri** [comment]
Some users are reporting problems in the wrong category, which causes the problem to get reported to the wrong body (the municipality or the national road agency). Because of this, we are thinking about the possibility to create some kind of mechanism where the municipalities can move a problem to the national road agency, and thus creating a new problem report e-mail (and probably notifying the user that the problem has been forwarded to the right agency as well).

This would probably take place on the main site (rather than in the administration control panel) as this is where the municipality users are working with FixaMinGata, and would probably require the user to be logged in to an account that represents the particular municipality.

We assume that this kind of functionality would depend on this issue being fixed.

Any thoughts on this would be very welcome!

**Comment by jonkri** [comment]
(I updated the above comment slightly.)

**Comment by dracos** [comment]
That sounds like a good idea, yes; the initial report to the body could contain a link saying "Has this been sent to the wrong authority?" or similar and then they could have a web page where they could redirect to a different body/category covering the same spot (I guess this would need more thought if we wanted to allow the body to change but keep the initial category, which should be theoretically possible I imagine!).

Something very similar that might be useful would be a report very close to a boundary, and the body realises the user has got the pin wrong slightly and it should be sent to the neighbouring body; a same interface for a recipient moving a report to a different body (potentially different category also there).

The link in the email could work immediately, or be tied to a body user, as you suggest.

**Comment by abimysoc** [comment]
Would be useful from a UK user support perspective too; we get the occasional email saying "my report has gone to the wrong body, can you re-send it to the right one please?" and at the moment we have to tell them sorry, we can't do that, and advise them to create a new report correcting whatever it was that caused it to go to the wrong one.

**Comment by dracos** [comment]
In that instance; Zurich admin has the ability to move a pin to reseat it, so if we got that in the main admin (it would need stuff from this ticket because of body/category updates I imagine), accompanied with the 'resend' button would fix the issue if they had placed it just on the wrong side of a border, for example.

**Comment by jonkri** [comment]
(Somewhat relevant to the above: We just got a list of reports that were on the border between two municipalities that had been sent to the wrong body. The municipality that fixed the issues wanted the issues to be assigned to them, presumably for statistics reasons. As a quick fix, I updated "bodies_str" in the "problem" table to the ID of the right municipality. Unfortunately, the text on the problem page which describes when the report was sent now is now incorrect, in that it mentions the wrong municipality.)

**Comment by dracos** [comment]
If you updated `bodies_str`, then the front end should then update to match, as that gets the information from that column live. The only time I can see that it wouldn't would be if the "external body" column of the report was set to something, in which case that would be used by preference. Can you check, as something odd must be going on if so!

**Comment by jonkri** [comment]
@dracos: I apologize for being unclear here. The municipality shown <strong>is</strong> the one that I changed to when I modify "bodies_str", so that part is correct. The problem is that the report wasn't sent to that municipality at the time that was mentioned; the report was sent to the "original" municipality, and it was not until I changed "bodies_str" and informed the right municipality that the report was "sent" to the municipality that is now mentioned on the page. I guess I just wanted to point out that particular (minor) shortcoming/abstraction leak in just updating "bodies_str".

**Comment by jonkri** [comment]
All-in-all, it would be nice to have a more "sophisticated" solution to the problem - one that allows us to see how the report has been moved between categories and bodies.

**Comment by dracos** [comment]
On your first bit, I guess you could click the "resend" report in the admin (or set whensent to null) so the report actually does get resent to the new body and the front end will then display that information correctly (or if you've informed then manually, update whensent manually for the same result).

**Comment by jonkri** [comment]
Thanks for this! Does this mean that it's also possible to change the body in the admin?

**Comment by dracos** [comment]
If you update the category of a report, its body will also be updated if necessary to the matching body. So yes, by changing the category, but not independently.
